lib/reek/configuration/app_configuration.rb

```diff
@@ -5,89 +5,108 @@ module Reek
-      EXCLUDE_PATHS_KEY     = 'exclude_paths'
-      @configuration        = {}
-      @has_been_initialized = false
+    class AppConfiguration
+      EXCLUDE_PATHS_KEY = 'exclude_paths'
+      attr_reader :exclude_paths, :default_directive, :directory_directives
 
-      class << self
-        attr_reader :configuration
+      # Given this configuration file:
+      #
+      # ---
+      # IrresponsibleModule:
+      #   enabled: false
+      # "app/helpers":
+      #   UtilityFunction:
+      #     enabled: false
+      # exclude_paths:
+      #   - "app/controllers"
+      #
+      # this would result in the following configuration:
+      #
+      # exclude_paths = [Pathname('spec/samples/two_smelly_files')]
+      # default_directive = { Reek::Smells::IrresponsibleModule=>{"enabled"=>false} }
+      # directory_directives = { Pathname("spec/samples/three_clean_files/") =>
+      #                          {Reek::Smells::UtilityFunction=>{"enabled"=>false}}}
+      #
+      #
+      # @param options [OpenStruct]
+      #   e.g.: #<OpenStruct report_format=:text,
+      #                      location_format=:numbers,
+      #                      colored=true,
+      #                      smells_to_detect=[],
+      #                      config_file=#<Pathname:config/defaults.reek>,
+      #                      argv=["lib/reek/spec"]>
+      def initialize(options = nil)
+        self.directory_directives = {}
+        self.default_directive    = {}
+        self.exclude_paths        = []
 
-        def initialize_with(options)
-          @has_been_initialized = true
-          configuration_file_path = ConfigurationFileFinder.find(options: options)
-          return unless configuration_file_path
-          load_from_file configuration_file_path
-        end
-
-        def configure_smell_repository(smell_repository)
-          # Let users call this method directly without having initialized AppConfiguration before
-          # and if they do, initialize it without application context
-          initialize_with(nil) unless @has_been_initialized
-          for_smell_types.each do |klass_name, config|
-            klass = load_smell_type(klass_name)
-            smell_repository.configure(klass, config) if klass
-          end
-        end
+        load options
+      end
 
-        def load_from_file(path)
-          if path.size.zero?
-            report_problem('Empty file', path)
-            return
-          end
+      # @param source_via [String] - the source of the code inspected.
+      # Might be a string, STDIN or Filename / Pathname. We're only interested in the source
+      # when it's coming from file since a directory_directive doesn't make sense
+      # for anything else.
+      # @return [Hash | nil] the configuration for the source or nil
+      def directory_directive_for_source(source_via)
+        return unless source_via
+        return unless Source::SourceCode.source_from_file?(source_via)
```

Comment:
We can just leave this out I think? The code below will be just fine if `"stdin"` or `"string"` is passed in.